### PR TITLE
Macpie/ch3562/fix nat lease renew

### PR DESCRIPTION
**Pull request by macpie**

### CODE REVIEW

File: test/proxy_SUITE.erl [hunk 1]
```diff
@@ -402,11 +411,20 @@ limit_exceeded(_Config) ->
     ok = libp2p_swarm:stop(ClientSwarm2),
 
     %% check we didn't leak any sockets here
-    ok = test_util:wait_until(fun() ->
-                                      [{_ID, Info}] = ranch:info(),
-                                      0 == proplists:get_value(active_connections, Info) andalso
-                                      0 == proplists:get_value(all_connections, Info)
-                              end),
+    ok = test_util:wait_until(
```
Comment (madninja) [inline test/proxy_SUITE.erl]: Repeating blocks of the same wait_until code? Extract out into it’s own function?

File: test/proxy_SUITE.erl [hunk 2]
```diff
@@ -402,11 +411,20 @@ limit_exceeded(_Config) ->
...
+        fun() ->
+            lists:foldl(
+                fun({_ID, _Info}, false) ->
+                    false;
+                ({_ID, Info}, _Acc) ->
+                    0 == proplists:get_value(active_connections, Info) andalso
```
Comment (madninja) [inline test/proxy_SUITE.erl]: Use assertions here

File: src/libp2p_transport_tcp.erl [hunk 3]
```diff
@@ -695,7 +695,7 @@ tcp_listen_addrs(Socket) ->
 
 reuseport0(TID, Type, {0,0,0,0}, 0) ->
     Cache = libp2p_swarm:cache(TID),
-    case libp2p_cache:lookup(Cache, {tcp_listen_addrs, Type}, []) of
+    case libp2p_cache:lookup(Cache, {tcp_local_listen_addrs, Type}, []) of
```
Comment (madninja) [inline src/libp2p_transport_tcp.erl]: This rename doesn’t fix anything right? Just makes it clearer what is being cached?

File: src/nat/libp2p_nat.erl [hunk 4]
```diff
@@ -180,7 +187,9 @@ retry_matrix(Try, Port) ->
 %% @end
 %%--------------------------------------------------------------------
 increment_port(Port) when Port < 65535-1 ->
-    Port+1.
+    Port+1;
+increment_port(_Port) ->
+    65535.
```
Comment (madninja) [inline src/nat/libp2p_nat.erl]: Just return Port here? Less hardcoding of value?

File: src/libp2p_transport_tcp.erl (hunk 3 above)
Comment (macpie) [reply src/libp2p_transport_tcp.erl]: yes

File: src/nat/libp2p_nat.erl (hunk 4 above)
Comment (macpie) [reply src/nat/libp2p_nat.erl]: indeed

File: test/proxy_SUITE.erl (hunk 2 above)
Comment (macpie) [reply test/proxy_SUITE.erl]: I can't use them here cause it's in the wait_until

File: src/nat/libp2p_nat_statem.erl [hunk 5]
```diff
@@ -69,20 +74,20 @@ terminate(_Reason, _State, _Data) ->
...
+started(cast, {register, Port, Lease, Since}, #data{tid=TID}=Data) ->
+    ok = update_cache(TID, Port),
+    ok = renew(Lease),
     {next_state, active, Data#data{port=Port, lease=Lease, since=Since}};
 started(Type, Content, Data) ->
     handle_event(Type, Content, Data).
 
-active(info, renew, #data{port=Port}=Data) ->
+active(info, renew, #data{port=Port, tid=TID}=Data) ->
     case libp2p_nat:add_port_mapping(Port, false) of
```
Comment (Vagabond) [inline src/nat/libp2p_nat_statem.erl]: The old map does not need to be released first?

File: src/nat/libp2p_nat_statem.erl [hunk 6]
```diff
@@ -100,8 +105,15 @@ handle_event(_Type, _Content, Data) ->
 %% Internal Function Definitions
 %% ------------------------------------------------------------------
 
+update_cache(TID, Port) ->
+    Cache = libp2p_swarm:cache(TID),
+    ok = libp2p_cache:insert(Cache, ?CACHE_KEY, Port).
```
Comment (Vagabond) [inline src/nat/libp2p_nat_statem.erl]: should the external IP be in here, too?

File: src/libp2p_cache.erl [hunk 7]
```diff
@@ -76,21 +80,22 @@ init([TID]) ->
     SwarmName = libp2p_swarm:name(TID),
     DataDir = libp2p_config:base_dir(TID),
     Opts = [{file, filename:join([DataDir, SwarmName, "cache.dets"])}],
-    {ok, Name} = dets:open_file(SwarmName, Opts),
-    {ok, #state{dets=Name}}.
+    {ok, Dets} = dets:open_file(SwarmName, Opts),
+    self() ! migrate,
```
[the remark this answers is not in the document]
Comment (macpie) [reply src/libp2p_cache.erl]: will do sync

File: src/nat/libp2p_nat_statem.erl [hunk 8]
```diff
@@ -50,13 +52,16 @@ start(Args) ->
...
+    ?CACHE_KEY.
+
 %% ------------------------------------------------------------------
 %% gen_statem Function Definitions
 %% ------------------------------------------------------------------
-init([Pid]=_Args) ->
+init([Pid, TID]=_Args) ->
     lager:info("init with ~p", [_Args]),
     true = erlang:link(Pid),
-    {ok, started, #data{}}.
```
[the remark this answers is not in the document]
Comment (macpie) [reply src/nat/libp2p_nat_statem.erl]: old leases are released before that

File: src/nat/libp2p_nat_statem.erl (hunk 6 above)
Comment (macpie) [reply src/nat/libp2p_nat_statem.erl]: we do no use it for nat

**Pull request merged**